src/webservice/WebService.h

```diff
@@ -7,49 +7,57 @@
 #ifndef WEBSERVICE_WEBSERVICE_H_
 #define WEBSERVICE_WEBSERVICE_H_
 
-#include "base/Base.h"
 #include "base/Status.h"
-#include <proxygen/httpserver/HTTPServer.h>
-#include "thread/NamedThread.h"
 
 DECLARE_int32(ws_http_port);
 DECLARE_int32(ws_h2_port);
 DECLARE_string(ws_ip);
 DECLARE_int32(ws_threads);
 
+namespace proxygen {
+class HTTPServer;
+class RequestHandler;
+}   // namespace proxygen
+
 namespace nebula {
+namespace thread {
+class NamedThread;
+}   // namespace thread
 
-using HandlerGen = std::unordered_map<
-    std::string,
-    std::function<proxygen::RequestHandler*()>>;
+namespace web {
+class Router;
+}   // namespace web
 
 class WebService final {
 public:
+    explicit WebService(const std::string& name = "");
+    ~WebService();
+
+    web::Router& router() {
+        CHECK(!started_) << "Don't add routes after starting web server!";
+        return *router_;
+    }
+
     // To start the global web server.
     // Two ports would be bound, one for HTTP, another one for HTTP2.
     // If FLAGS_ws_http_port or FLAGS_ws_h2_port is zero, an ephemeral port
     // would be assigned and set back to the gflag, respectively.
-    static MUST_USE_RESULT Status start();
-    // To stop the web service and join the internal threads
-    static void stop();
-
-    // To register a handler generator for a specific path
-    // All registrations have to be done before calling start().
-    // Anything registered after start() being called will not take
-    // effect
-    //
-    // By default, web service will register the handler for getting/setting
-    // the flags, getting the stats
-    static void registerHandler(const std::string& path,
-                                std::function<proxygen::RequestHandler*()>&& gen);
+    MUST_USE_RESULT Status start();
 
-private:
-    WebService() = delete;
+    // To stop the web service and join the internal threads
+    void stop();
 
-    static std::unique_ptr<proxygen::HTTPServer> server_;
-    static std::unique_ptr<thread::NamedThread> wsThread_;
+    // Check whether this web service is stopped
+    bool stopped() const {
+        return stopped_;
+    }
 
-    static HandlerGen handlerGenMap_;
+private:
+    bool started_{false};
```

Comment:
For such a scenario, we typically use either an atomic boolean or a volatile one.

BTW, is it really necessary to use two flags?

Reply:
OK, stopped member is not necessary now, I have removed it. Since the web service only used in one thread, so I don't prevent the accession to the variable from other threads.